### PR TITLE
Soulbound NFTs

### PR DESCRIPTION
**Pull request by Rishi556**
Soulbound NFT's will be tied to a user. They many not transfer it to others or place it for sale on the market or auction it. Soulbound will need to be specified at creation time and cannot be changed.

### CODE REVIEW

**Comment by Rishi556** [comment]
Finished manual testing, need to write automated tests

File: contracts/nft.js [hunk 1]
```diff
@@ -1573,6 +1570,7 @@ actions.issue = async (payload) => {
             if (finalLockNfts.length > 0) {
               newInstance = {
                 account: finalTo,
+                soulBound,
```
Comment (bt-cryptomancer) [inline contracts/nft.js:1573]: If soulBound is not explicitly defined in the issue payload, then I'd prefer it to be defaulted to false here (or else not added at all to the newInstance object).

File: contracts/nft.js [hunk 2]
```diff
@@ -1581,6 +1579,7 @@ actions.issue = async (payload) => {
             } else {
               newInstance = {
                 account: finalTo,
+                soulBound,
```
Comment (bt-cryptomancer) [inline contracts/nft.js:1582]: same thing here, if soulBound is not explicitly defined in the issue payload, then I'd prefer it to be defaulted to false (or else not added at all to the newInstance object).

File: contracts/nft.js (hunk 1 above)
Comment (Rishi556) [reply contracts/nft.js]: Changed in https://github.com/hive-engine/hivesmartcontracts/pull/2/commits/f2d8c42c1ddce3aaeccf57d03d17305386bdca90

File: contracts/nft.js (hunk 2 above)
Comment (Rishi556) [reply contracts/nft.js]: Changed in Changed in https://github.com/hive-engine/hivesmartcontracts/pull/2/commits/f2d8c42c1ddce3aaeccf57d03d17305386bdca90

**Comment by bt-cryptomancer** [comment]
Overall this looks pretty good, and you've got the right idea with the test cases (I've skimmed over them but didn't check each line in detail).  Do you have a test case to verify that you can't lock soulbound NFTs inside other NFTs (using the lockNfts input to the issue action)?

Also, I notice the validation checks that you added to the transfer and delegate actions were not added to the burn action. Was this an oversight, or do you think it's fine to allow soulbound NFTs to be burned?

Please also add to the test cases for the nftmarket, nftauction, and nftairdrops contracts to ensure you can't sell, auction, or airdrop soulbound NFTs.  Simple checks added to the existing test cases should be fine, no need to go overboard here.

For the nftairdrops contract specifically, I think you will need to modify the `parseAndValidateAirdrop` function so that validation fails if any of the NFTs are soulbound.  It should be sufficient to modify this query and error message:

image

I don't think any other contracts need modification, as they indirectly rely on the NFT transfer action for validation.

**Comment by Rishi556** [comment]
> I notice the validation checks that you added to the transfer and delegate actions were not added to the burn action. Was this an oversight, or do you think it's fine to allow soulbound NFTs to be burned?

A couple developers mentioned that they'd like the ability for users to be able to burn soulbound nfts. I think that's alright since it allows users a way to get rid of them in the event they want to fully remove connections to something.

> Do you have a test case to verify that you can't lock soulbound NFTs inside other NFTs (using the lockNfts input to the issue action)?

Not yet, but will make sure all current tests in the nft contract will have a soulbound version tested.

I'll be sure to test all the other contracts which rely on nfts. I agree that not too much needs to be tested on the 3 you mentioned. I'm imagining testing to make sure non soulbound can be placed on the market, and soulbound can't, and same with auctions and the rest.

Appreciate your feedback.

**Comment by bt-cryptomancer** [comment]
> A couple developers mentioned that they'd like the ability for users to be able to burn soulbound nfts. I think that's alright since it allows users a way to get rid of them in the event they want to fully remove connections to something.

Okay I agree, don't see any potential issues with allowing this.

> I'll be sure to test all the other contracts which rely on nfts. I agree that not too much needs to be tested on the 3 you mentioned. I'm imagining testing to make sure non soulbound can be placed on the market, and soulbound can't, and same with auctions and the rest.

Yep, that should cover it.  Plus the minor code change that I mentioned to the nftairdrops contract.

**Review by bt-cryptomancer — APPROVED** [review]
Everything looks good now, thanks!  Will look to get this released soon.

**Pull request merged**